Mathlib/CategoryTheory/Shift/Adjunction.lean

```diff
@@ -0,0 +1,447 @@
+
+Once we have established all this, the compatibility of the commutation isomorphism for
+`F` expressed in `CommShift.zero` and `CommShift.add` immediately implies the similar
+statements for the commutation isomorphisms for `G`.
+
+TODO: Construct a `CommShift` structure on `F` from a `CommShift` structure on `G`, using
+opposite categories.
+
+-/
+
+namespace CategoryTheory
+
+open Category
+
+namespace Adjunction
+
+variable {C D : Type*} [Category C] [Category D]
+  {F : C ⥤ D} {G : D ⥤ C} (adj : F ⊣ G) {A : Type*} [AddMonoid A] [HasShift C A] [HasShift D A]
+
+namespace CommShift
+
+variable {a b : A} (e₁ : shiftFunctor C a ⋙ F ≅ F ⋙ shiftFunctor D a)
+    (e₁' : shiftFunctor C a ⋙ F ≅ F ⋙ shiftFunctor D a)
+    (f₁ : shiftFunctor C b ⋙ F ≅ F ⋙ shiftFunctor D b)
+    (e₂ : shiftFunctor D a ⋙ G ≅ G ⋙ shiftFunctor C a)
+    (e₂' : shiftFunctor D a ⋙ G ≅ G ⋙ shiftFunctor C a)
+    (f₂ : shiftFunctor D b ⋙ G ≅ G ⋙ shiftFunctor C b)
+
+/-- Given an adjunction `adj : F ⊣ G`, `a` in `A` and commutation isomorphisms
+`e₁ : shiftFunctor C a ⋙ F ≅ F ⋙ shiftFunctor D a` and
+`e₂ : shiftFunctor D a ⋙ G ≅ G ⋙ shiftFunctor C a`, this expresses the compatibility of
+`e₁` and `e₂` with the unit of the adjunction `adj`.
+-/
+abbrev CompatibilityUnit :=
+  ∀ (X : C), (adj.unit.app X)⟦a⟧' = adj.unit.app (X⟦a⟧) ≫ G.map (e₁.hom.app X) ≫ e₂.hom.app _
+
+/-- Given an adjunction `adj : F ⊣ G`, `a` in `A` and commutation isomorphisms
+`e₁ : shiftFunctor C a ⋙ F ≅ F ⋙ shiftFunctor D a` and
+`e₂ : shiftFunctor D a ⋙ G ≅ G ⋙ shiftFunctor C a`, this expresses the compatibility of
+`e₁` and `e₂` with the counit of the adjunction `adj`.
+-/
+abbrev CompatibilityCounit :=
+  ∀ (Y : D), adj.counit.app (Y⟦a⟧) = F.map (e₂.hom.app Y) ≫ e₁.hom.app _ ≫ (adj.counit.app Y)⟦a⟧'
+
+/-- Given an adjunction `adj : F ⊣ G`, `a` in `A` and commutation isomorphisms
+`e₁ : shiftFunctor C a ⋙ F ≅ F ⋙ shiftFunctor D a` and
+`e₂ : shiftFunctor D a ⋙ G ≅ G ⋙ shiftFunctor C a`, compatibility of `e₁` and `e₂` with the
+unit of the adjunction `adj` implies compatibility with the counit of `adj`.
+-/
+lemma compatibilityCounit_of_compatibilityUnit (h : CompatibilityUnit adj e₁ e₂) :
+    CompatibilityCounit adj e₁ e₂ := by
+  intro Y
+  have eq := h (G.obj Y)
+  rw [← cancel_mono (e₂.inv.app _ ≫ G.map (e₁.inv.app _))] at eq
+  dsimp at eq ⊢
+  conv_rhs at eq => slice 3 4; rw [Iso.hom_inv_id_app]
+  conv_rhs at eq => erw [id_comp]; rw [← Functor.map_comp, Iso.hom_inv_id_app];
+                    erw [Functor.map_id]; rw [comp_id]
+  apply (adj.homEquiv _ _).injective
+  rw [adj.homEquiv_unit, adj.homEquiv_unit, G.map_comp, adj.unit_naturality_assoc, ← eq]
+  slice_rhs 4 5 => rw [← Functor.map_comp, ← assoc, Iso.inv_hom_id_app]; erw [id_comp]
+  erw [← e₂.inv.naturality]; rw [← assoc _ _ (e₂.inv.app Y), Functor.comp_map, ← Functor.map_comp]
+  simp only [Functor.id_obj, Functor.comp_obj, right_triangle_components, Functor.map_id, id_comp,
+    Iso.hom_inv_id_app]
```

Comment:
In most situations, it is advisable not to do `rw [← assoc]`, but to always apply it forward, and use `_assoc` versions of lemmas:
```suggestion
  simp only [← cancel_mono (e₂.inv.app _ ≫ G.map (e₁.inv.app _)),
    Category.assoc, Category.assoc, Iso.hom_inv_id_app_assoc, comp_id,
    ← Functor.map_comp, Iso.hom_inv_id_app, Functor.comp_obj, Functor.map_id] at eq
  apply (adj.homEquiv _ _).injective
  dsimp
  rw [adj.homEquiv_unit, adj.homEquiv_unit, G.map_comp, adj.unit_naturality_assoc, ← eq]
  simp only [Category.assoc, ← Functor.map_comp, Iso.inv_hom_id_app_assoc]
  erw [← e₂.inv.naturality]
  dsimp
  simp only [right_triangle_components, ← Functor.map_comp_assoc, Functor.map_id, id_comp,
    Iso.hom_inv_id_app, Functor.comp_obj]
```